pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -176,6 +195,29 @@ func (ctx rollingUpgradeCtx) run() ([]corev1.Pod, error) {
 	return deletedPods, nil
 }
 
+// isNonHACluster returns true if the expected and actual number of master nodes indicates that the quorum of that cluster
+// does not allow the loss of any node in which case a regular rolling upgrade might not be possible especially when doing
+// a major version upgrade.
+func isNonHACluster(actualPods []corev1.Pod, expectedMasters []string) bool {
+	if len(expectedMasters) > 2 {
+		return false
+	}
+	actualMasters := label.FilterMasterNodePods(actualPods)
+	return len(actualMasters) <= 2
+}
+
+func isMajorVersionUpgrade(es esv1.Elasticsearch) (bool, error) {
+	specVersion, err := version.Parse(es.Spec.Version)
+	if err != nil {
+		return false, err
+	}
+	statusVersion, err := version.Parse(es.Status.Version)
```

Comment:
This creates a dependency on the resource's `Status`, I'm wondering if we should consider the situation where `es.Status.Version` is an empty string. My feeling is that `es.Status.Version` is managed on a "[best-effort](https://github.com/elastic/cloud-on-k8s/blob/96006dec0977d9fdb84b8be2a41bd9c4c03d2670/pkg/controller/elasticsearch/reconcile/state.go#L103)" basis, and the upgrade process should not be prevented because the status is not fully up-to-date yet. If we cannot parse `es.Status.Version` maybe we should err on the safe side and assume that this is a regular update.

Reply:
I understand your argument. But we are already basing upgrade decisions on the existence of this value in the association mechanism where a upgrade is not rolled out if that version is the empty string or the version indicates that an update is still pending for the associated resource. So it feels to me like we are already committed to keeping this up to date in a relatively timely manner and if anything we should make sure that these status updates succeed. 

In addition if we were to fall back to a rolling upgrade, there is a 50% chance that the upgrade will not go through and we/the user have a broken cluster afterwards.